### PR TITLE
improvement(matrix): Use recommended eslint config 

### PR DESCRIPTION
**Pull request by jzaffiro**
Updates the eslint config from minimal-deprecated to recommended, and addresses the resulting violations.

[AB#3013](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/3013)

### CODE REVIEW

File: packages/dds/matrix/src/test/matrix.applyStashedOp.spec.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@
  * Licensed under the MIT License.
  */
 
-import { strict as assert } from "assert";
+import { strict as assert } from "node:assert";
```
Comment (jzaffiro) [inline packages/dds/matrix/src/test/matrix.applyStashedOp.spec.ts:6]: This was autofixed in all of the test files due to [unicorn/prefer-node-protocol](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/v48.0.1/docs/rules/prefer-node-protocol.md), but the current variation violates [import/no-nodejs-modules](https://github.com/import-js/eslint-plugin-import/blob/v2.29.0/docs/rules/no-nodejs-modules.md). Not sure which error is better to supress, or if there's a way around it.

File: packages/dds/matrix/.eslintrc.cjs [hunk 2]
```diff
@@ -22,7 +19,7 @@ module.exports = {
 			files: ["*.spec.ts", "src/test/**"],
 			rules: {
 				// Test files are run in node only so additional node libraries can be used.
-				"import/no-nodejs-modules": ["error", { allow: ["assert", "path"] }],
+				"import/no-nodejs-modules": ["off", { allow: ["assert", "path"] }],
```
Comment (jzaffiro) [inline packages/dds/matrix/.eslintrc.cjs:22]: Turned off since many of the test files use assert, and some use path. I left a comment in `matrix.applyStashedOp.spec.ts` about this as well - if this isn't the right fix, let me know.

File: packages/dds/matrix/src/sparsearray2d.ts [hunk 3]
```diff
@@ -43,7 +43,7 @@ const r0c0ToMorton2x16 = (row: number, col: number): number =>
 	(r0ToMorton16(row) | c0ToMorton16(col)) >>> 0;
 
 type RecurArrayHelper<T> = RecurArray<T> | T;
-type RecurArray<T> = RecurArrayHelper<T>[];
+export type RecurArray<T> = RecurArrayHelper<T>[];
```
Comment (jzaffiro) [inline packages/dds/matrix/src/sparsearray2d.ts:46]: Used in `matrix.ts` - is there any reason that this shouldn't be exported?

File: packages/dds/matrix/src/test/matrix.applyStashedOp.spec.ts (hunk 1 above)
Comment (Josmithr) [reply packages/dds/matrix/src/test/matrix.applyStashedOp.spec.ts:6]: You should be able to update the package's eslintrc to fix this. It should already have an `no-nodejs-modules` override that specifies `assert` - you'll just want to update that to say `node:assert` instead.

File: packages/dds/matrix/.eslintrc.cjs (hunk 2 above)
Comment (Josmithr) [reply packages/dds/matrix/.eslintrc.cjs:22]: You should be able to update this to
```suggestion
				"import/no-nodejs-modules": ["error", { allow: ["node:assert", "node:path"] }],
```
to get things back in a good state

File: packages/dds/matrix/src/handlecache.ts [hunk 4]
```diff
@@ -67,8 +69,10 @@ export class HandleCache implements IVectorConsumer<Handle> {
 		}
 	}
 
-	/** Used by 'CacheMiss()' to retrieve handles for a range of positions. */
-	private getHandles(start: number, end: number) {
+	/**
+	 * Used by 'CacheMiss()' to retrieve handles for a range of positions.
```
Comment (Josmithr) [inline packages/dds/matrix/src/handlecache.ts:73]:
```suggestion
	 * Used by {@link HandleCache.cacheMiss} to retrieve handles for a range of positions.
```

File: packages/dds/matrix/src/handlecache.ts [hunk 5]
```diff
@@ -40,7 +40,7 @@ export class HandleCache implements IVectorConsumer<Handle> {
 	 * Throws a 'RangeError' if the provided 'position' is out-of-bounds wrt. the
```
Comment (Josmithr) [inline packages/dds/matrix/src/handlecache.ts:40]:
```suggestion
	 * @throws A 'RangeError' if the provided 'position' is out-of-bounds with regards to the
```

File: packages/dds/matrix/src/permutationvector.ts [hunk 6]
```diff
@@ -378,22 +384,25 @@ export class PermutationVector extends Client {
...
 
-	private readonly onMaintenance = (args: IMergeTreeMaintenanceCallbackArgs) => {
+	private readonly onMaintenance = (args: IMergeTreeMaintenanceCallbackArgs): void => {
 		if (args.operation === MergeTreeMaintenanceType.UNLINK) {
 			let freed: number[] = [];
 
 			for (const { segment } of args.deltaSegments) {
 				const asPerm = segment as PermutationSegment;
 				if (isHandleValid(asPerm.start)) {
-					// Note: Using the spread operator with `.splice()` can exhaust the stack.
```
Comment (Josmithr) [inline packages/dds/matrix/src/permutationvector.ts:393]: Given this comment, I wonder if we want to revert the change here and just disable the rule.

File: packages/dds/matrix/src/sparsearray2d.ts (hunk 3 above)
Comment (Josmithr) [reply packages/dds/matrix/src/sparsearray2d.ts:46]: Should be fine IMO - we're not exporting it from the package.

File: packages/dds/matrix/src/sparsearray2d.ts [hunk 7]
```diff
@@ -83,12 +87,14 @@ export class SparseArray2D<T>
 		return undefined;
 	}
 
-	public get matrixProducer() {
-		// eslint-disable-next-line @typescript-eslint/no-unsafe-return
+	public get matrixProducer(): IMatrixProducer<T> {
+		// Suppression needed to satisfy the interface declaration while also not returning
```
Comment (Josmithr) [inline packages/dds/matrix/src/sparsearray2d.ts:91]: Could we return `undefined as unknown as IMatrixProducer<T>`? Would remove the need for the liner overrides, though I think the comment you added would be great to keep.

File: packages/dds/matrix/src/sparsearray2d.ts [hunk 8]
```diff
@@ -219,17 +235,19 @@ export class SparseArray2D<T>
...
-		// eslint-disable-next-line @typescript-eslint/no-unsafe-return
-		return level ?? (parent[subKey] = new Array(256).fill(undefined));
+		// Using new Array is needed because the array created with Array.from does not
+		// satisfy (T|undefined)[].
+		// eslint-disable-next-line unicorn/no-new-array
+		return level ?? (parent[subKey] = new Array<T|undefined>(256).fill(undefined));
 	}
 
-	public snapshot() {
+	public snapshot(): UA<UA<UA<UA<UA<T>>>>> {
```
Comment (Josmithr) [inline packages/dds/matrix/src/sparsearray2d.ts:246]: Whoa. This is funky. I don't know this code well, but just to sanity check: is this implementing an interface method with simpler signature?

File: packages/dds/matrix/src/permutationvector.ts (hunk 6 above)
Comment (jzaffiro) [reply packages/dds/matrix/src/permutationvector.ts:393]: I was wondering this too, but since we don't use splice anymore, I didn't know if the comment was still relevant or not. I can definitely revert and disable though if we want.

File: packages/dds/matrix/src/sparsearray2d.ts (hunk 8 above)
Comment (jzaffiro) [reply packages/dds/matrix/src/sparsearray2d.ts:246]: No, unfortunately not :(. It follows the typing of the root, which was defined ~4 years ago (presumably when the file was created).

File: packages/dds/matrix/src/permutationvector.ts (hunk 6 above)
Comment (Josmithr) [reply packages/dds/matrix/src/permutationvector.ts:393]: Just to be safe, that might be best.

File: packages/dds/matrix/src/sparsearray2d.ts (hunk 8 above)
Comment (Josmithr) [reply packages/dds/matrix/src/sparsearray2d.ts:245]: OOF. Okay, well if nothing else, this maybe exposes some previously hidden code smell.

**Review by Josmithr — APPROVED** [review]

**Review by anthony-murphy — APPROVED** [review]
